src/security/waf_remote_cfg.cpp

```diff
@@ -0,0 +1,931 @@
+    cur_appsec_cfg_.user_config_remove_config(key);
+  }
+
+ private:
+  CurrentAppSecConfig &cur_appsec_cfg_;
+};
+
+class ConfigurationEndListener : public rc::Listener {
+ public:
+  ConfigurationEndListener(std::function<void()> func)
+      : func_{std::move(func)} {}
+
+  rc::Products get_products() override { return {}; }
+
+  rc::Capabilities get_capabilities() override { return {}; }
+
+  dd::Optional<std::string> on_update(const Configuration &) override {
+    return {};
+  };
+
+  void on_revert(const Configuration &config) override {}
+
+  void on_post_process() override { func_(); }
+
+ private:
+  std::function<void()> func_;
+};
+
+class AppSecConfigService {
+  std::shared_ptr<dnsec::ddwaf_owned_map> default_config_;
+  CurrentAppSecConfig current_config_;
+  std::shared_ptr<datadog::tracing::Logger> logger_;
+
+  static inline std::unique_ptr<AppSecConfigService> instance_;  // NOLINT
+
+  AppSecConfigService(dnsec::ddwaf_owned_map default_config,
+                      std::shared_ptr<datadog::tracing::Logger> logger)
+      : default_config_{std::make_shared<dnsec::ddwaf_owned_map>(
+            std::move(default_config))},
+        logger_{std::move(logger)} {
+    current_config_.set_dd_config(this->default_config_);
+  }
+
+ public:
+  AppSecConfigService(const AppSecConfigService &) = delete;
+  AppSecConfigService &operator=(const AppSecConfigService &) = delete;
+  AppSecConfigService(AppSecConfigService &&) = delete;
+  AppSecConfigService &operator=(AppSecConfigService &&) = delete;
+  ~AppSecConfigService() = default;
+
+  static void initialize(dnsec::ddwaf_owned_map default_config,
+                         std::shared_ptr<datadog::tracing::Logger> logger) {
+    if (instance_) {
+      throw std::logic_error{"AppSecConfigService already initialized"};
+    }
+    instance_ = std::unique_ptr<AppSecConfigService>{
+        new AppSecConfigService{std::move(default_config), std::move(logger)}};
+  }
+
+  static bool has_instance() { return static_cast<bool>(instance_); }
+
+  static AppSecConfigService &instance() {
+    if (!instance_) {
+      throw std::logic_error{"AppSecConfigService not initialized"};
+    }
+    return *instance_;
+  }
+
+  void subscribe_to_remote_config(datadog::tracing::DatadogAgentConfig &ddac,
+                                  bool accept_cfg_update,
+                                  bool is_subscribe_activation) {
+    if (is_subscribe_activation) {
+      subscribe_activation(ddac);
+    }
+
+    if (accept_cfg_update) {
+      subscribe_rules_and_data(ddac);
+
+      ddac.remote_configuration_listeners.emplace_back(
+          new ConfigurationEndListener([this] {
+            std::optional<dnsec::ddwaf_owned_map> maybe_upd =
+                current_config_.merged_update_config();
+            if (maybe_upd) {
+              auto &&upd = *maybe_upd;
+              dnsec::Library::update_ruleset(upd.get());
+            }
+          }));
+    }
```

Comment:
I mean, dd-trace-cpp, when applying the configuration, already does a dispatch to the listener based on the product(s) subscribed. I can just take advantage of that facility rather than implement my own dispatch logic (probably twice, for on_update and on_revert). The extra memory usage is minimal, likely under 150 bytes for the two extra listeners plus some pointers, and the the cpu is likely better (avoid dispatch based product twice) but doesn't matter since remote config application is not a hot stop. I did optimize the listeners a bit by removing some virtual calls and storage of product and capability in the listeners with CRTP.

Reply:
As long as you're aware of the consequences, that's fine. However, `on_post_process` is called for all listeners, but, the right behaviour should be: only listeners that were previously notified should be called by `on_post_process`.  My reasoning is, there's no purpose in notifying a listener if you weren't notified in the first place.

Reply:
That is not an accurate description of the current code in dd-trace-cpp; post process is called unconditionally on all listeners; see https://github.com/DataDog/dd-trace-cpp/blob/main/src/datadog/remote_config/remote_config.cpp#L285-L287

As to whether it's a good idea to change it or not... IMO on_post_process should be a on a separate class of listeners as it was on my design -- its point was to coordinate changes across multiple listeners (not just multiple configurations for the same listener) and most product listeners won't need it anyway.

Reply:
My bad, I guess I was not explicit enough. I wasn't telling how `on_post_process` works today but how it **should** (and will) work.